src/transformStateWithClones.js

```diff
@@ -7,7 +7,37 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const statesArray = [];
+
+  let currentState = { ...state };
+
+  actions.forEach((action) => {
+    switch (action.type) {
+      case 'clear':
+        currentState = {};
+        break;
+
+      case 'addProperties':
+        currentState = { ...currentState, ...action.extraData };
+        break;
+
+      case 'removeProperties':
+        const copy = { ...currentState };
+        action.keysToRemove.forEach((key) => {
+          delete copy[key];
+        });
+
+        currentState = { ...copy };
+        break;
+
+      default:
+        break;
+    }
+
+    statesArray.push({ ...currentState });
```

Comment:
You are pushing a copy of `currentState` into `statesArray` after each action. This is correct to avoid mutations in the subsequent actions. However, consider that this operation is common to all cases and could be moved outside of the switch statement to avoid repetition.